R/integrated_scores.R

```diff
@@ -26,13 +26,18 @@ weighted_survival_score = function(loss, truth, distribution, times, t_max, p_ma
     unique_times = .c_get_unique_times(truth[, "time"], times)
   }
 
-  is_distr_or_array =
-    (inherits(distribution, "Distribution")) ||
-    (inherits(distribution, "array") & length(dim(array)) == 3)
-
-  if (is_distr_or_array) {
+  # get the cdf
+  if (inherits(distribution, "Distribution")) {
     cdf = as.matrix(distribution$cdf(unique_times))
-  } else {
+  }
+  else if (inherits(distribution, "array") &
+           length(dim(distribution)) == 3) {
+    # 'distribution' is a survival 3d array so create an
+    # `Arrdist` using the 'median' curve
+    arrdistr = distr6::as.Distribution(1 - distribution, fun = "cdf",
+      decorators = c("CoreStatistics", "ExoticStatistics"))
+    cdf = as.matrix(arrdistr$cdf(unique_times))
+  } else { # 'distribution' is a survival 2d matrix
```

Comment:
@bblodfon  The existing code deliberately avoided coercing back to distributions as this was a massive bottleneck, we *really* want to avoid this. Instead I suggest an internal function in proba to coerce the array to a matrix using whatever the selected curve was in the `predict` function -- you could possibly call some of the internal distr6 functions to do this

Reply:
Please check the new code optimization commits. I cannot pass the `which.curve` parameter in the `score()` so I use the 'median' survival matrix, as is documented in `PredictionSurv`. I used the `.ext_surv_mat()` which is also used in `.surv_return()`, so no conversion to `Arrdist` is done.